general/kdtree.hpp

```diff
@@ -0,0 +1,765 @@
+            RNS(pt,R, itb+siz/2+1, ite, level, res, dist); // right
+            RNS(pt,R, itb, itb+siz/2, level,   res, dist); // left
+
+            // check central one
+            Tfloat dd=Dist(mtb->pt, pt);
+            if (dd<R)
+            {
+               res.push_back(mtb->ind);
+               dist.push_back(dd);
+            }
+         }
+      }
+      else
+      {
+         Tfloat dd;
+         for (auto it=itb; it!=ite; it++)
+         {
+            dd=Dist(it->pt, pt);
+            if (dd<R) // update bc
+            {
+               res.push_back(it->ind);
+               dist.push_back(dd);
+            }
+         }
+      }
+   }
+
+
+};
+
+
+/// Defines KDTree in 3D
+typedef KDTree<int,double,3> KDTree3D;
+
+/// Defines Contructs KDTree in 2D
+typedef KDTree<int,double,2> KDTree2D;
+
+/// Defines KDTree in 1D
+typedef KDTree<int,double,1> KDTree1D;
+
+
+/// Defines KDTree with points stored outside the class.
+/// The external storage slows down the search and the construction
+/// methods.
+class KDTreeM
+{
+public:
+   /// Construtor
+   KDTreeM(mfem::Vector& coords_,int dim_)
+   {
+      gf=&coords_;
+      ndim=dim_;
+      tp.SetSize(ndim);
+      int np=(gf->Size())/ndim;
+      data.resize(np);
+      for (int i=0; i<np; i++)
+      {
+         data[i].xx=coords_.GetData()+i*ndim;
+         data[i].ind=i;
+      }
+      Sort();
+   }
+
+   /// Internal structure used as a node in the KDTree
+   struct NodeND
+   {
+      size_t ind;
+      double *xx;//coordinates of the node
+   };
+
+
+   /// Returns the spatial dimension of the points
+   int SpatialDimension() {return ndim;}
+
+   /// Data iterator
+   typedef typename std::vector<NodeND>::iterator iterator;
+
+   /// Returns iterator to beginning of the point cloud
+   iterator begin() {return data.begin();}
+
+   /// Returns iterator to the end of the point cloud
+   iterator end() {return data.end();}
+
+   /// Returns the size of the point cloud
+   size_t size() {return data.size();}
+
+   /// Clears the point cloud.
+   void clear() { data.clear();}
+
+   /// Builds the KDTree.
+   void Sort() { SortInPlace(data.begin(),data.end(),0);}
+
+   /// Returns the index of the closest point to the input point pt.
+   int FindClosestPoint(mfem::Vector& pt);
+
+   ///  Returns the closest point and the distance to the input point pt.
+   void FindClosestPoint(Vector& pt, int &ind, double& dist);
+
+   /// Brute force search - please, use it only for debuging purposes.
+   void _FindClosestPoint(Vector& pt,int& ind, double& dist);
+
+private:
+   mfem::Vector* gf;
+   int ndim;
+   Vector tp;
+
+   /// The actual data in the KDTree.
+   std::vector<NodeND> data;
+
+   /// Structure utilized for nearest neighbor search (NNS)
+   struct PointS
+   {
+      double dist;
+      size_t pos;
+      size_t level;
+      NodeND sp; //index of the point
+   };
+
+   /// Functor utilized in the coordinate comparison
+   /// for building the KDTree
+   struct CompN
```

Comment:
Same question as above - why not an std::function?

Reply:
That is a very specific comparison function which has a state std::uint8_t dim;
It compares only the dim^th coordinate of a point.